File: lc/A003_Longest_Substring_Without_Repeating_Characters.js


var lengthOfLongestSubstring = (s) => {
    let res = 0, from = 0;
    let m = new Map();

    for (let i = 0; i < s.length; i++) {
        if (m.has(s.charAt(i))) from = Math.max(from, m.get(s.charAt(i) ) + 1);
        // if (m.has(s.charAt(i))) from = m.get(s.charAt(i) ) + 1;
        m.set(s.charAt(i), i);
        res = Math.max(res, i - from + 1);
    }
    return res;
}

const lls2 = (s) => {
    let map = new Array(128).fill(0);
    let begin = 0, end = 0, counter = 0, maxLength = 0;
    
    while (end < s.length) {
        if (map[s.charAt(end++).charCodeAt(0)]++ > 0) counter++;
        while (counter > 0) {
            if (map[s.charAt(begin++).charCodeAt(0)]-- > 1) counter--;
        }
        maxLength = Math.max(maxLength, end - begin);
    }
    
    return maxLength;
}

console.log(lengthOfLongestSubstring('abba'));

File: java/src/leetcode/problems/A003_Longest_Substring_Without_Repeating_Characters.java
package leetcode.problems;

public class A003_Longest_Substring_Without_Repeating_Characters {
    private int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int[] freq = new int[128];
        int count = 0;
        int bp = 0, ep = 0, l = 1;
        while (ep < s.length()) {
            if (freq[s.charAt(ep++)]++ > 0) count++;
            while (count > 0) {
                if (freq[s.charAt(bp++)]-- > 1) count--;
            }
            l = Math.max(ep - bp, l);
        }
        return l;
    }
}
